File: utils/handlers/clans.js
const { Scenes, Markup } = require('telegraf');
const clanService = require('../../services/clanService');
const db = require('../../database');
const ui = require('../ui');
const media = require('../media');

/**
 * Clan Scene: The Syndicate Hub
 */
const clanScene = new Scenes.BaseScene('clan_scene');

clanScene.enter(async (ctx) => {
    const user = await db.users.findOne({ telegramId: ctx.from.id });
    if (!user.school) {
        return ctx.replyWithHTML("🏯 <b>ACADEMY REQUIRED</b>\nYou must choose your path (Tokyo vs Kyoto) before joining a syndicate.\n\nUse /school to decide.");
    }
    
    if (!user.clanId) {
        return clanScene.renderJoinOrCreate(ctx, user);
    }
    return clanScene.renderClanHub(ctx, user);
});

clanScene.renderJoinOrCreate = async function(ctx, user) {
    let msg = ui.formatHeader("CLAN RECRUITMENT") + "\n\n" +
        "You are not part of a syndicate. Join a clan to share resources and compete globally.\n\n" +
        "🔓 <b>Min Level to Join:</b> 15\n" +
        "🏗 <b>Min Level to Create:</b> 20 (Costs 5,000 G)";

    const kb = Markup.inlineKeyboard([
        [Markup.button.callback('🔍 Browse Clans', 'browse_clans')],
        [Markup.button.callback('🏗 Create Syndicate (5k G)', 'init_create_clan')],
        [Markup.button.callback('🔙 Return to Hub', 'back_to_hub')]
    ]);

    if (ctx.callbackQuery) {
        await media.smartEdit(ctx, msg, kb);
    } else {
        await ctx.replyWithHTML(msg, kb);
    }
};

clanScene.renderClanHub = async function(ctx, user) {
    const clan = await db.clans.findOne({ _id: user.clanId });
    if (!clan) return module.exports.renderJoinOrCreate(ctx, user);

    let msg = ui.formatHeader(`SYNDICATE: ${clan.name}`) + "\n\n" +
        `🏫 <b>Academy:</b> ${clan.school}\n` +
        `🏷 <b>Tag:</b> [${clan.tag}]\n` +
        `👥 <b>Members:</b> ${clan.members.length}/${clan.slots || 25}\n` +
        `📈 <b>Power Score:</b> <code>${clan.totalElo || 0}</code>\n` +
        `✨ <b>Treasury:</b> <code>${clan.treasury.dust || 0}</code> Dust\n\n` +
        `🏆 <b>Your Role:</b> ${user.clanRole}\n`;

    const kb = [
        [Markup.button.callback('💎 Contribute Dust', 'contribute_dust')],
        [Markup.button.callback('🚪 Leave Syndicate', 'leave_clan')]
    ];

    if (user.clanRole === 'Leader') {
        kb.push([Markup.button.callback('🏗 Expand (+5 Slots, 5k G)', 'expand_clan')]);
        kb.push([Markup.button.callback('⚙️ Leader Panel', 'leader_panel')]);
    }

    kb.push([Markup.button.callback('🔙 Return to Hub', 'back_to_hub')]);

    if (ctx.callbackQuery) {
        await media.smartEdit(ctx, msg, Markup.inlineKeyboard(kb));
    } else {
        await ctx.replyWithHTML(msg, Markup.inlineKeyboard(kb));
    }
};

clanScene.action('init_create_clan', async (ctx) => {
    const user = await db.users.findOne({ telegramId: ctx.from.id });
    if (user.playerLevel < 20) return ctx.answerCbQuery("❌ Level 20 required to create!", { show_alert: true });
    
    await ctx.answerCbQuery();
    return ctx.reply("Please use the following format to create a clan:\n/create_clan NAME TAG\nExample: /create_clan 'Kyoto High' KYT");
});

clanScene.action('leave_clan', async (ctx) => {
    const result = await clanService.leaveClan(ctx.from.id);
    if (result.success) return clanScene.enter(ctx);
});

clanScene.action('expand_clan', async (ctx) => {
    const res = await clanService.expandClan(ctx.from.id);
    await ctx.answerCbQuery(res.msg, { show_alert: !res.success });
    if (res.success) {
        const user = await db.users.findOne({ telegramId: ctx.from.id });
        return clanScene.renderClanHub(ctx, user);
    }
});

clanScene.action('browse_clans', async (ctx) => {
    const user = await db.users.findOne({ telegramId: ctx.from.id });
    const clans = await db.clans.find({ school: user.school }).sort({ totalElo: -1 }).limit(10);
    
    if (clans.length === 0) return ctx.answerCbQuery("No clans found in your Academy.", { show_alert: true });

    let msg = ui.formatHeader(`DISCOVER: ${user.school.toUpperCase()}`) + "\n\n";
    clans.forEach((c, i) => {
        msg += `<b>#${i+1} [${c.tag}] ${c.name}</b>\n└ 👥 ${c.members.length}/${c.slots || 25} | 📈 ${c.totalElo}\n\n`;
    });

    const kb = clans.map(c => [Markup.button.callback(`Join ${c.tag}`, `join_clan_${c._id}`)]);
    kb.push([Markup.button.callback('⬅️ Back', 'back_to_hub')]);

    return media.smartEdit(ctx, msg, Markup.inlineKeyboard(kb));
});

clanScene.action(/join_clan_(.+)/, async (ctx) => {
    const res = await clanService.joinClan(ctx.from.id, ctx.match[1]);
    await ctx.answerCbQuery(res.msg, { show_alert: !res.success });
    if (res.success) {
        return clanScene.enter(ctx);
    }
});

// Clan specific listeners for scene actions
clanScene.action('back_to_hub', (ctx) => {
    ctx.answerCbQuery().catch(() => null);
    return ctx.scene.leave();
});

// Listener for the global "clan_home" callback
const showHome = async (ctx) => {
    return ctx.scene.enter('clan_scene');
};

module.exports = { clanScene, showHome };

File: utils/handlers/clans.py
from aiogram import Router, types, F
from aiogram.utils.keyboard import InlineKeyboardBuilder
from aiogram.filters import Command

from database import db
from utils import ui, media
from services.clan_service import clan_service

router = Router()

@router.message(Command("clan", "syndicate"))
@router.callback_query(F.data == "cmd_clan")
@router.callback_query(F.data == "clan_home")
async def cmd_clan(callback_or_message: types.CallbackQuery | types.Message, user: dict):
    if not user.get('school'):
        msg = "🏯 <b>ACADEMY REQUIRED</b>\nYou must choose your path (Tokyo vs Kyoto) before joining a syndicate.\n\nUse /school to decide."
        if isinstance(callback_or_message, types.CallbackQuery):
            return await callback_or_message.answer(msg, show_alert=True)
        return await callback_or_message.answer(msg, parse_mode='HTML')

    if isinstance(callback_or_message, types.CallbackQuery):
        await callback_or_message.answer()

    if not user.get('clanId'):
        return await render_join_or_create(callback_or_message, user)
    return await render_clan_hub(callback_or_message, user)

async def render_join_or_create(callback_or_message, user):
    msg = ui.format_header("CLAN RECRUITMENT") + "\n\n" + \
          "You are not part of a syndicate. Join a clan to share resources and compete globally.\n\n" + \
          "🔓 <b>Min Level to Join:</b> 15\n" + \
          "🏗 <b>Min Level to Create:</b> 20 (Costs 5,000 G)"

    builder = InlineKeyboardBuilder()
    builder.row(types.InlineKeyboardButton(text="🔍 Browse Clans", callback_data="browse_clans"))
    builder.row(types.InlineKeyboardButton(text="🏗 Create Syndicate (5k G)", callback_data="init_create_clan"))
    builder.row(types.InlineKeyboardButton(text="🔙 Return to Hub", callback_data="back_to_hub"))

    if isinstance(callback_or_message, types.CallbackQuery):
        await media.smart_edit(callback_or_message.message, msg, reply_markup=builder.as_markup())
    else:
        await callback_or_message.answer(msg, parse_mode='HTML', reply_markup=builder.as_markup())

async def render_clan_hub(callback_or_message, user):
    clan = await db.clans.find_one({"_id": user['clanId']})
    if not clan:
        return await render_join_or_create(callback_or_message, user)

    msg = ui.format_header(f"SYNDICATE: {clan['name']}") + "\n\n" + \
          f"🏫 <b>Academy:</b> {clan['school']}\n" + \
          f"🏷 <b>Tag:</b> [{clan['tag']}]\n" + \
          f"👥 <b>Members:</b> {len(clan['members'])}/{clan.get('slots', 25)}\n" + \
          f"📈 <b>Power Score:</b> <code>{clan.get('totalElo', 0)}</code>\n" + \
          f"✨ <b>Treasury:</b> <code>{clan['treasury'].get('dust', 0)}</code> Dust\n\n" + \
          f"🏆 <b>Your Role:</b> {user.get('clanRole', 'Member')}\n"

    builder = InlineKeyboardBuilder()
    builder.row(types.InlineKeyboardButton(text="💎 Contribute Dust", callback_data="contribute_dust"))
    builder.row(types.InlineKeyboardButton(text="🚪 Leave Syndicate", callback_data="leave_clan"))

    if user.get('clanRole') == 'Leader':
        builder.row(types.InlineKeyboardButton(text="🏗 Expand (+5 Slots, 5k G)", callback_data="expand_clan"))
        builder.row(types.InlineKeyboardButton(text="⚙️ Leader Panel", callback_data="leader_panel"))

    builder.row(types.InlineKeyboardButton(text="🔙 Return to Hub", callback_data="back_to_hub"))

    if isinstance(callback_or_message, types.CallbackQuery):
        await media.smart_edit(callback_or_message.message, msg, reply_markup=builder.as_markup())
    else:
        await callback_or_message.answer(msg, parse_mode='HTML', reply_markup=builder.as_markup())

@router.callback_query(F.data == "init_create_clan")
async def handle_init_create(callback: types.CallbackQuery, user: dict):
    if user.get('playerLevel', 1) < 20:
        return await callback.answer("❌ Level 20 required to create!", show_alert=True)
    
    await callback.answer()
    await callback.message.reply("Please use the following format to create a clan:\n/create_clan NAME TAG\nExample: /create_clan 'Kyoto High' KYT")

@router.message(Command("create_clan"))
async def handle_create_clan(message: types.Message, user: dict):
    args = message.text.split(maxsplit=2)
    if len(args) < 3:
        return await message.reply("❌ <b>USAGE:</b> /create_clan [NAME] [TAG]\nExample: /create_clan 'Jujutsu High' JJH", parse_mode='HTML')
    
    name = args[1]
    tag = args[2][:4]
    
    result = await clan_service.create_clan(message.from_user.id, name, tag)
    if not result['success']:
        return await message.reply(f"❌ {result['msg']}")
    
    await message.reply(f"✅ {result['msg']}")
    user.update({"clanId": result['clan']['_id'], "clanRole": "Leader"})
    await render_clan_hub(message, user)

@router.callback_query(F.data == "browse_clans")
async def handle_browse_clans(callback: types.CallbackQuery, user: dict):
    clans = await db.clans.find({"school": user['school']})
    # Sort by totalElo desc in code or query
    clans.sort(key=lambda x: x.get('totalElo', 0), reverse=True)
    clans = clans[:10]
    
    if not clans:
        return await callback.answer("No clans found in your Academy.", show_alert=True)

    msg = ui.format_header(f"DISCOVER: {user['school'].upper()}") + "\n\n"
    builder = InlineKeyboardBuilder()
    for i, c in enumerate(clans):
        msg += f"<b>#{i+1} [{c['tag']}] {c['name']}</b>\n└ 👥 {len(c['members'])}/{c.get('slots', 25)} | 📈 {c.get('totalElo', 0)}\n\n"
        builder.row(types.InlineKeyboardButton(text=f"Join {c['tag']}", callback_data=f"join_clan_{c['_id']}"))

    builder.row(types.InlineKeyboardButton(text="⬅️ Back", callback_data="cmd_clan"))
    await callback.answer()
    await media.smart_edit(callback.message, msg, reply_markup=builder.as_markup())

@router.callback_query(F.data.startswith("join_clan_"))
async def handle_join_clan(callback: types.CallbackQuery, user: dict):
    clan_id = callback.data.replace("join_clan_", "")
    result = await clan_service.join_clan(callback.from_user.id, clan_id)
    
    if not result['success']:
        return await callback.answer(result['msg'], show_alert=True)
    
    await callback.answer(result['msg'], show_alert=True)
    user.update({"clanId": clan_id, "clanRole": "Member"})
    await render_clan_hub(callback, user)

@router.callback_query(F.data == "leave_clan")
async def handle_leave_clan(callback: types.CallbackQuery, user: dict):
    result = await clan_service.leave_clan(callback.from_user.id)
    if not result['success']:
        return await callback.answer(result['msg'], show_alert=True)
    
    await callback.answer(result['msg'], show_alert=True)
    user.update({"clanId": None, "clanRole": None})
    await render_join_or_create(callback, user)

@router.callback_query(F.data == "expand_clan")
async def handle_expand_clan(callback: types.CallbackQuery, user: dict):
    result = await clan_service.expand_clan(callback.from_user.id)
    await callback.answer(result['msg'], show_alert=not result['success'])
    if result['success']:
        await render_clan_hub(callback, user)
